grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcServiceBuilder.java

```diff
@@ -326,6 +331,80 @@ public GrpcServiceBuilder addService(String path, BindableService bindableServic
         return this;
     }
 
+    /**
+     * Adds a gRPC {@link BindableService} with functions to this {@link GrpcServiceBuilder}.
+     * The functions will be applied in the specified order and before functions decorated with the service.
+     * In the following sample code, these functions will be applied
+     * in the order of FirstDecorator, SecondDecorator, ClassDecorator, and MethodDecorator.
+     * <pre>{@code
+     * @Decorator(ClassDecorator.class)
+     * class SimpleServiceImpl extends SimpleServiceImplBase {
+     *
+     *     @Decorator(MethodDecorator.class)
+     *     @Override
+     *     public void simpleCall(SimpleRequest request,
+     *                            StreamObserver<SimpleResponse> responseObserver) {
+     *         ...
+     *     }
+     * }
+     * ...
+     * GrpcService.builder()
+     *            .addService(new SimpleServiceImpl(),
+     *                        List.of(delegate -> delegate.decorate(new FirstDecorator()),
+     *                                delegate -> delegate.decorate(new SecondDecorator())))
+     *            .build()
+     * }</pre>
+     */
+    @UnstableApi
+    public GrpcServiceBuilder addService(
+            BindableService bindableService,
+            Iterable<? extends Function<? super HttpService, ? extends HttpService>> decorators) {
+        requireNonNull(bindableService, "bindableService");
+        requireNonNull(decorators, "decorators");
+        final ServerServiceDefinition serverServiceDefinition = bindableService.bindService();
+        registryBuilder.addService(serverServiceDefinition, bindableService.getClass());
+        additionalDecorators.put(serverServiceDefinition, decorators);
+        return this;
+    }
+
+    /**
+     * Adds an implementation of gRPC service with functions to this {@link GrpcServiceBuilder}.
+     * @see #addService(Object, Function)
+     * @see #addService(BindableService, Iterable)
+     */
+    @UnstableApi
+    public <T> GrpcServiceBuilder addService(
+            T implementation,
+            Function<? super T, ServerServiceDefinition> serviceDefinitionFactory,
+            Iterable<? extends Function<? super HttpService, ? extends HttpService>> decorators) {
+        requireNonNull(implementation, "implementation");
+        requireNonNull(serviceDefinitionFactory, "serviceDefinitionFactory");
+        requireNonNull(decorators, "decorators");
+        final ServerServiceDefinition serverServiceDefinition = serviceDefinitionFactory.apply(implementation);
+        requireNonNull(serverServiceDefinition, "serviceDefinitionFactory.apply() returned null");
+        registryBuilder.addService(serverServiceDefinition, implementation.getClass());
+        additionalDecorators.put(serverServiceDefinition, decorators);
```

Comment:
How about adding the `decorators` when we call `registryBuilder.addService` instead of managing separately using `additionalDecorators`?
If so, I think we can gather adding and applying decorators into the same logic. 🤔

Reply:
According to the https://github.com/line/armeria/pull/4202, we must keep annotated decorators as `DecoratorAndOrder` to Inject dependencies in  `GrpcDecoratingService`.

> If so, I think we can gather adding and applying decorators into the same logic

So, I think we could not do it because we need to keep separately annotated decorators and dynamically added decorators.
If we call `registryBuilder.addService` instead of managing separately using `additionalDecorators`, eventually we need to keep them separately and merge two decorators in `GrpcDecoratingService#serviceAdded`.
There seem to be no differences compared to the current implementation other than the location where we keep `additionalDecorators`.

Reply:
> we must keep annotated decorators as DecoratorAndOrder

That's true. We need separate lists for that.

> There seem to be no differences compared to the current implementation other than the location where we keep additionalDecorators.

That's exactly. 😄 We can keep the decorators as is.
However, I thought if we keep it inside the `HandlerRegistry` we don't need this logic anymore:
https://github.com/line/armeria/pull/4316/files#diff-bdf258d37f85fff3213fdd0725deebed224916d18dcc3e290e3914ce77a9312fR902

Also, I thought we need to rename `HandlerRegistry.decorators()` to something like `annotationDecorators` because it only returns some of the decorators.

By the way, are we adding this variant?
```java
public GrpcServiceBuilder addService(
        String path, BindableService bindableService,
        MethodDescriptor<?, ?> methodDescriptor,
        Iterable<? extends Function<? super HttpService, ? extends HttpService>> decorators) {
```
If so, we need to change the key of the map to `ServerMethodDefinition`.

Reply:
> However, I thought if we keep it inside the HandlerRegistry we don't need this logic anymore:

It seems better. 😄
I will address this. also the above variant should be added.